File: Two_Sum/Solution1.java
/* 
	Solution 1 
	
	Insert all the values of the array into a hashmap for quick index lookup.
	Loop through the array using the current value to search the hashmap for a possible missing value.
		missing value = target - current value.
	If found, return the current index and use the map to get the missing value's index.
*/


/* 
	Time Complexity: O(n)
	Space Complexity: O(n)
*/


import java.util.Arrays;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        // answer
        int[] result = {-1,-1};
       
        // create a map to handle index searching
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < nums.length; i++){
            map.put(nums[i], i);
        }
        
        // loop through array finding a pair that sums
        for(int i = 0; i < nums.length; i++){
            int missing = target - nums[i];
            if(map.containsKey(missing)){
                if(map.get(missing) != i){
                    result[0] = i;
                    result[1] = map.get(missing);
                    return result;
                }
            }
        }

        // ERROR RESULT
        return result;
        
    }
    
}

/*
	LEETCODE RESULTS

	Runtime: 3 ms, faster than 62.36% of Java online submissions for Two Sum.
	Memory Usage: 40.4 MB, less than 17.58% of Java online submissions for Two Sum.
*/

File: Destination_City/Solution1.cs
/* 
	Solution 1 
	
	Enter in each path into a hashmap<start, destination>. Then pick any path to start at and follow
	until you reach a city that is not in the hashmap.
*/


/* 
	Time Complexity: O(n) time, where n is the number of paths
	Space Complexity: O(n)
*/

public class Solution {
    public string DestCity(IList<IList<string>> paths) {
        // HashMap for quick lookup
        Dictionary<string, string> route = new Dictionary<string, string>();
        
        // Enter in paths to hashmap
        foreach(List<string> path in paths){
            route.Add(path[0], path[1]);
        }
        
        // Traverse path starting from paths[0]
        string city = paths[0][0];
        while(route.ContainsKey(city)){
            city = route[city];
        }
        
        // Final destination
        return city;
    }
}

/*
	LEET CODE RESULTS

	Runtime: 104 ms, faster than 99.22% of C# online submissions for Destination City.
	Memory Usage: 27.4 MB, less than 45.74% of C# online submissions for Destination City.
*/